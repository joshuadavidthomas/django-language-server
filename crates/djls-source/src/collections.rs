use dashmap::{DashMap, DashSet};
use rustc_hash::FxHasher;
use std::hash::BuildHasherDefault;

pub type FxDashMap<K, V> = DashMap<K, V, BuildHasherDefault<FxHasher>>;
pub type FxDashSet<K> = DashSet<K, BuildHasherDefault<FxHasher>>;
