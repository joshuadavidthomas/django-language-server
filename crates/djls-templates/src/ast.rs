use serde::Serialize;
use thiserror::Error;

use crate::tokens::Token;
use crate::tokens::TokenStream;
use crate::tokens::TokenType;

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize)]
pub struct Ast {
    nodelist: Vec<Node>,
    line_offsets: LineOffsets,
}

impl Ast {
    #[must_use]
    pub fn nodelist(&self) -> &Vec<Node> {
        &self.nodelist
    }

    #[must_use]
    pub fn line_offsets(&self) -> &LineOffsets {
        &self.line_offsets
    }

    pub fn add_node(&mut self, node: Node) {
        self.nodelist.push(node);
    }

    pub fn set_line_offsets(&mut self, tokens: &TokenStream) {
        for token in tokens.tokens() {
            if let TokenType::Newline = token.token_type() {
                if let Some(start) = token.start() {
                    // Add offset for next line
                    self.line_offsets.add_line(start + 1);
                }
            }
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub struct LineOffsets(pub Vec<u32>);

impl LineOffsets {
    /// Create a new LineOffsets from source text
    #[must_use]
    pub fn new(text: &str) -> Self {
        let mut offsets = Self::default();
        let mut pos = 0;
        for ch in text.chars() {
            if ch == '\n' {
                offsets.add_line((pos + 1) as u32);
            }
            pos += ch.len_utf8();
        }
        offsets
    }

    pub fn add_line(&mut self, offset: u32) {
        self.0.push(offset);
    }

    #[must_use]
    pub fn position_to_line_col(&self, position: usize) -> (usize, usize) {
        let position = u32::try_from(position).unwrap_or_default();
        let line = match self.0.binary_search(&position) {
            Ok(exact_line) => exact_line,    // Position is at start of this line
            Err(0) => 0,                     // Before first line start
            Err(next_line) => next_line - 1, // We're on the previous line
        };

        // Calculate column as offset from line start
        let col = if line == 0 {
            position as usize
        } else {
            (position - self.0[line]) as usize
        };

        // Convert to 1-based line number
        (line + 1, col)
    }

    #[must_use]
    pub fn line_col_to_position(&self, line: u32, col: u32) -> u32 {
        // line is 1-based, so subtract 1 to get the index
        self.0[(line - 1) as usize] + col
    }
}

impl Default for LineOffsets {
    fn default() -> Self {
        Self(vec![0])
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub enum Node {
    Tag {
        name: String,
        bits: Vec<String>,
        span: Span,
    },
    Comment {
        content: String,
        span: Span,
    },
    Text {
        content: String,
        span: Span,
    },
    Variable {
        var: String,
        filters: Vec<String>,
        span: Span,
    },
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
pub struct Span {
    start: u32,
    length: u32,
}

impl Span {
    #[must_use]
    pub fn new(start: u32, length: u32) -> Self {
        Self { start, length }
    }

    #[allow(clippy::trivially_copy_pass_by_ref)]
    #[must_use]
    pub fn start(&self) -> u32 {
        self.start
    }

    #[allow(clippy::trivially_copy_pass_by_ref)]
    #[must_use]
    pub fn length(&self) -> u32 {
        self.length
    }
}

impl From<Token> for Span {
    fn from(token: Token) -> Self {
        let start = token.start().unwrap_or(0);
        let length = u32::try_from(token.content().len()).unwrap_or(0);
        Span::new(start, length)
    }
}

#[derive(Clone, Debug, Error, PartialEq, Eq, Serialize)]
pub enum AstError {
    #[error("Empty AST")]
    EmptyAst,
    #[error("Invalid tag '{tag}' structure: {reason}")]
    InvalidTagStructure {
        tag: String,
        reason: String,
        span: Span,
    },
    #[error("Unbalanced structure: '{opening_tag}' at {opening_span:?} missing closing '{expected_closing}'")]
    UnbalancedStructure {
        opening_tag: String,
        expected_closing: String,
        opening_span: Span,
        closing_span: Option<Span>,
    },
    #[error("Invalid {node_type} node: {reason}")]
    InvalidNode {
        node_type: String,
        reason: String,
        span: Span,
    },
    #[error("Unclosed tag: {tag}")]
    UnclosedTag { tag: String, span: Span },
    #[error("Orphaned tag '{tag}' - {context}")]
    OrphanedTag {
        tag: String,
        context: String,
        span: Span,
    },
    #[error("endblock '{name}' does not match any open block")]
    UnmatchedBlockName { name: String, span: Span },
    #[error("Tag '{tag}' requires at least {min} argument{}", if *.min == 1 { "" } else { "s" })]
    MissingRequiredArguments { 
        tag: String, 
        min: usize,
        span: Span 
    },
    #[error("Tag '{tag}' accepts at most {max} argument{}", if *.max == 1 { "" } else { "s" })]
    TooManyArguments {
        tag: String,
        max: usize,
        span: Span
    },
}

impl AstError {
    /// Get the span of this error, if available
    #[must_use]
    pub fn span(&self) -> Option<Span> {
        match self {
            AstError::EmptyAst => None,
            AstError::InvalidTagStructure { span, .. } => Some(*span),
            AstError::UnbalancedStructure { opening_span, .. } => Some(*opening_span),
            AstError::InvalidNode { span, .. } => Some(*span),
            AstError::UnclosedTag { span, .. } => Some(*span),
            AstError::OrphanedTag { span, .. } => Some(*span),
            AstError::UnmatchedBlockName { span, .. } => Some(*span),
            AstError::MissingRequiredArguments { span, .. } => Some(*span),
            AstError::TooManyArguments { span, .. } => Some(*span),
        }
    }

    /// Get a numeric error code for this error type
    #[must_use]
    pub fn error_code(&self) -> u16 {
        match self {
            AstError::EmptyAst => 1,
            AstError::InvalidTagStructure { .. } => 2,
            AstError::UnbalancedStructure { .. } => 3,
            AstError::InvalidNode { .. } => 4,
            AstError::UnclosedTag { .. } => 5,
            AstError::OrphanedTag { .. } => 6,
            AstError::UnmatchedBlockName { .. } => 7,
            AstError::MissingRequiredArguments { .. } => 8,
            AstError::TooManyArguments { .. } => 9,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;
    use crate::parser::Parser;

    mod line_offsets {
        use super::*;

        #[test]
        fn test_new_starts_at_zero() {
            let offsets = LineOffsets::default();
            assert_eq!(offsets.position_to_line_col(0), (1, 0)); // Line 1, column 0
        }

        #[test]
        fn test_start_of_lines() {
            let mut offsets = LineOffsets::default();
            offsets.add_line(10); // Line 2 starts at offset 10
            offsets.add_line(25); // Line 3 starts at offset 25

            assert_eq!(offsets.position_to_line_col(0), (1, 0)); // Line 1, start
            assert_eq!(offsets.position_to_line_col(10), (2, 0)); // Line 2, start
            assert_eq!(offsets.position_to_line_col(25), (3, 0)); // Line 3, start
        }
    }

    mod spans_and_positions {
        use super::*;

        #[test]
        fn test_variable_spans() {
            let template = "Hello\n{{ user.name }}\nWorld";
            let tokens = Lexer::new(template).tokenize().unwrap();
            let mut parser = Parser::new(tokens);
            let (nodelist, errors) = parser.parse().unwrap();
            assert!(errors.is_empty());

            // Find the variable node
            let nodes = nodelist.nodelist();
            let var_node = nodes
                .iter()
                .find(|n| matches!(n, Node::Variable { .. }))
                .unwrap();

            if let Node::Variable { span, .. } = var_node {
                // Variable starts after newline + "{{"
                let (line, col) = nodelist
                    .line_offsets()
                    .position_to_line_col(span.start() as usize);
                assert_eq!(
                    (line, col),
                    (2, 0),
                    "Variable should start at line 2, col 3"
                );

                assert_eq!(span.length(), 9, "Variable span should cover 'user.name'");
            }
        }
    }
}
