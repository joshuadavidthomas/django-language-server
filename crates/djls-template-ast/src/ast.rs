use serde::Serialize;
use std::collections::BTreeMap;
use thiserror::Error;

#[derive(Clone, Debug, Default, Serialize)]
pub struct Ast {
    nodes: Vec<Node>,
    errors: Vec<AstError>,
}

impl Ast {
    pub fn nodes(&self) -> &Vec<Node> {
        &self.nodes
    }

    pub fn errors(&self) -> &Vec<AstError> {
        &self.errors
    }

    pub fn add_node(&mut self, node: Node) {
        self.nodes.push(node);
    }

    pub fn add_error(&mut self, error: AstError) {
        self.errors.push(error);
    }

    pub fn finalize(&mut self) -> Result<Ast, AstError> {
        if self.nodes.is_empty() && self.errors.is_empty() {
            return Err(AstError::EmptyAst);
        }
        Ok(self.clone())
    }
}

#[derive(Clone, Debug, Serialize)]
pub enum Node {
    Django(DjangoNode),
    Html(HtmlNode),
    Script(ScriptNode),
    Style(StyleNode),
    Text(String),
}

#[derive(Clone, Debug, Serialize)]
pub enum DjangoNode {
    Comment(String),
    Tag(TagNode),
    Variable {
        bits: Vec<String>,
        filters: Vec<DjangoFilter>,
    },
}

#[derive(Clone, Debug, Serialize)]
pub enum TagNode {
    Block {
        name: String,
        bits: Vec<String>,
        children: Vec<Node>,
    },
    Branch {
        name: String,
        bits: Vec<String>,
        children: Vec<Node>,
    },
    Closing {
        name: String,
        bits: Vec<String>,
    },
}

#[derive(Clone, Debug, Serialize)]
pub struct DjangoFilter {
    name: String,
    arguments: Vec<String>,
}

impl DjangoFilter {
    pub fn new(name: String, arguments: Vec<String>) -> Self {
        Self { name, arguments }
    }
}

#[derive(Clone, Debug, Serialize)]
pub enum HtmlNode {
    Comment(String),
    Doctype(String),
    Element {
        tag_name: String,
        attributes: Attributes,
        children: Vec<Node>,
    },
    Void {
        tag_name: String,
        attributes: Attributes,
    },
}

#[derive(Clone, Debug, Serialize)]
pub enum ScriptNode {
    Comment {
        content: String,
        kind: ScriptCommentKind,
    },
    Element {
        attributes: Attributes,
        children: Vec<Node>,
    },
}

#[derive(Clone, Debug, Serialize)]
pub enum ScriptCommentKind {
    SingleLine, // //
    MultiLine,  // /* */
}

#[derive(Clone, Debug, Serialize)]
pub enum StyleNode {
    Comment(String),
    Element {
        attributes: Attributes,
        children: Vec<Node>,
    },
}

#[derive(Clone, Debug, Serialize)]
pub enum AttributeValue {
    Value(String),
    Boolean,
}

pub type Attributes = BTreeMap<String, AttributeValue>;

#[derive(Clone, Debug, Error, Serialize)]
pub enum AstError {
    #[error("Empty AST")]
    EmptyAst,
    #[error("Stream error: {0}")]
    StreamError(String),
    #[error("Unclosed tag: {0}")]
    UnclosedTag(String),
    #[error("Unexpected tag: {0}")]
    UnexpectedTag(String),
    #[error("Invalid tag: {0}")]
    InvalidTag(String),
    #[error("Block error: {0} in {1}")]
    BlockError(String, String),
    #[error("Argument error: {0} - {1}")]
    ArgumentError(String, String),
    #[error("Unexpected token")]
    UnexpectedToken,
    #[error("Unexpected end of file")]
    UnexpectedEof,
}
