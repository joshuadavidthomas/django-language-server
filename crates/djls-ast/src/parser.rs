use crate::ast::{
    Ast, AstError, AttributeValue, DjangoFilter, DjangoNode, DjangoTagKind, HtmlNode, Node,
    ScriptCommentKind, ScriptNode, StyleNode,
};
use crate::tokens::{Token, TokenStream, TokenType};
use std::collections::BTreeMap;
use std::str::FromStr;
use thiserror::Error;

pub struct Parser {
    tokens: TokenStream,
    current: usize,
}

impl Parser {
    pub fn new(tokens: TokenStream) -> Self {
        Parser { tokens, current: 0 }
    }

    pub fn parse(&mut self) -> Result<Ast, ParserError> {
        let mut ast = Ast::default();

        while !self.is_at_end() {
            let node = self.next_node();
            match node {
                Ok(node) => {
                    ast.add_node(node);
                }
                Err(ParserError::AtEndOfStream) => {
                    if ast.nodes().is_empty() {
                        return Err(ParserError::UnexpectedEof);
                    }
                    break;
                }
                Err(_) => {
                    self.synchronize(&[
                        TokenType::DjangoBlock(String::new()),
                        TokenType::HtmlTagOpen(String::new()),
                        TokenType::HtmlTagVoid(String::new()),
                        TokenType::ScriptTagOpen(String::new()),
                        TokenType::StyleTagOpen(String::new()),
                        TokenType::Newline,
                        TokenType::Eof,
                    ])?;
                    continue;
                }
            }
        }

        Ok(ast.finalize()?)
    }

    fn next_node(&mut self) -> Result<Node, ParserError> {
        let token = self.peek()?;
        let node = match token.token_type() {
            TokenType::Comment(s, start, end) => self.parse_comment(s, start, end.as_deref()),
            TokenType::DjangoBlock(s) => self.parse_django_block(s),
            TokenType::DjangoVariable(s) => self.parse_django_variable(s),
            TokenType::Eof => self.parse_eof(),
            TokenType::HtmlTagClose(tag) => Err(ParserError::ClosingTagFound(tag.to_string())),
            TokenType::HtmlTagOpen(s) => self.parse_html_tag_open(s),
            TokenType::HtmlTagVoid(s) => self.parse_html_tag_void(s),
            TokenType::Newline => self.parse_newline(),
            TokenType::ScriptTagOpen(s) => self.parse_script_tag_open(s),
            TokenType::ScriptTagClose(_) => Err(ParserError::ClosingTagFound("script".to_string())),
            TokenType::StyleTagOpen(s) => self.parse_style_tag_open(s),
            TokenType::StyleTagClose(_) => Err(ParserError::ClosingTagFound("style".to_string())),
            TokenType::Text(s) => self.parse_text(s),
            TokenType::Whitespace(_) => self.parse_whitespace(),
        }?;
        Ok(node)
    }

    fn parse_comment(
        &mut self,
        content: &str,
        start: &str,
        end: Option<&str>,
    ) -> Result<Node, ParserError> {
        self.consume()?;

        match start {
            "{#" => Ok(Node::Django(DjangoNode::Comment(content.to_string()))),
            "<!--" => Ok(Node::Html(HtmlNode::Comment(content.to_string()))),
            "//" => Ok(Node::Script(ScriptNode::Comment {
                content: content.to_string(),
                kind: ScriptCommentKind::SingleLine,
            })),
            "/*" => {
                // Look back for script/style context
                let token_type = self
                    .peek_back(self.current)?
                    .iter()
                    .find_map(|token| match token.token_type() {
                        TokenType::ScriptTagOpen(_) => {
                            Some(TokenType::ScriptTagOpen(String::new()))
                        }
                        TokenType::StyleTagOpen(_) => Some(TokenType::StyleTagOpen(String::new())),
                        TokenType::ScriptTagClose(_) | TokenType::StyleTagClose(_) => None,
                        _ => None,
                    })
                    .ok_or(ParserError::InvalidMultLineComment)?;

                match token_type {
                    TokenType::ScriptTagOpen(_) => Ok(Node::Script(ScriptNode::Comment {
                        content: content.to_string(),
                        kind: ScriptCommentKind::MultiLine,
                    })),
                    TokenType::StyleTagOpen(_) => {
                        Ok(Node::Style(StyleNode::Comment(content.to_string())))
                    }
                    _ => unreachable!(),
                }
            }
            _ => Err(ParserError::UnexpectedToken(Token::new(
                TokenType::Comment(
                    content.to_string(),
                    start.to_string(),
                    end.map(String::from),
                ),
                0,
                None,
            ))),
        }
    }

    fn parse_django_block(&mut self, s: &str) -> Result<Node, ParserError> {
        self.consume()?;

        let bits: Vec<String> = s.split_whitespace().map(String::from).collect();
        let kind = DjangoTagKind::from_str(&bits[0])?;

        // If this is an end tag, signal it like we do with HTML closing tags
        if bits[0].starts_with("end") {
            return Err(ParserError::ClosingTagFound(bits[0].clone()));
        }

        let mut children = Vec::new();
        let end_tag = format!("end{}", bits[0]);

        while !self.is_at_end() {
            match self.next_node() {
                Ok(node) => {
                    println!("found django child node: {:?}", node);
                    children.push(node);
                }
                Err(ParserError::ClosingTagFound(tag)) => {
                    if tag == end_tag {
                        self.consume()?;
                        break;
                    }
                    // If it's not our end tag, keep collecting children
                }
                Err(e) => return Err(e),
            }
        }

        Ok(Node::Django(DjangoNode::Tag {
            kind,
            bits,
            children,
        }))
    }

    fn parse_django_variable(&mut self, s: &str) -> Result<Node, ParserError> {
        self.consume()?;

        let parts: Vec<&str> = s.split('|').collect();

        let bits: Vec<String> = parts[0].trim().split('.').map(String::from).collect();

        let filters: Vec<DjangoFilter> = parts[1..]
            .iter()
            .map(|filter_str| {
                let filter_parts: Vec<&str> = filter_str.trim().split(':').collect();
                let name = filter_parts[0].to_string();

                let arguments = if filter_parts.len() > 1 {
                    filter_parts[1]
                        .trim_matches('"')
                        .split(',')
                        .map(|arg| arg.trim().to_string())
                        .collect()
                } else {
                    Vec::new()
                };

                DjangoFilter::new(name, arguments)
            })
            .collect();

        Ok(Node::Django(DjangoNode::Variable { bits, filters }))
    }

    fn parse_eof(&mut self) -> Result<Node, ParserError> {
        if self.is_at_end() {
            self.consume()?;
            self.next_node()
        } else {
            Err(ParserError::UnexpectedEof)
        }
    }

    fn parse_html_tag_open(&mut self, s: &str) -> Result<Node, ParserError> {
        self.consume()?;

        let mut parts = s.split_whitespace();

        let tag_name = parts
            .next()
            .ok_or(ParserError::InvalidTokenAccess)?
            .to_string();

        let mut attributes = BTreeMap::new();

        for attr in parts {
            if let Some((key, value)) = attr.split_once('=') {
                // Key-value attribute (class="container")
                attributes.insert(
                    key.to_string(),
                    AttributeValue::Value(value.trim_matches('"').to_string()),
                );
            } else {
                // Boolean attribute (disabled)
                attributes.insert(attr.to_string(), AttributeValue::Boolean);
            }
        }

        let mut children = Vec::new();

        while !self.is_at_end() {
            match self.next_node() {
                Ok(node) => {
                    children.push(node);
                }
                Err(ParserError::ClosingTagFound(tag)) => {
                    if tag == tag_name {
                        self.consume()?;
                        break;
                    }
                }
                Err(e) => return Err(e),
            }
        }

        Ok(Node::Html(HtmlNode::Element {
            tag_name,
            attributes,
            children,
        }))
    }

    fn parse_html_tag_void(&mut self, s: &str) -> Result<Node, ParserError> {
        self.consume()?;
        let mut parts = s.split_whitespace();

        let tag_name = parts
            .next()
            .ok_or(ParserError::InvalidTokenAccess)?
            .to_string();

        let mut attributes = BTreeMap::new();

        for attr in parts {
            if let Some((key, value)) = attr.split_once('=') {
                attributes.insert(
                    key.to_string(),
                    AttributeValue::Value(value.trim_matches('"').to_string()),
                );
            } else {
                attributes.insert(attr.to_string(), AttributeValue::Boolean);
            }
        }

        Ok(Node::Html(HtmlNode::Void {
            tag_name,
            attributes,
        }))
    }

    fn parse_newline(&mut self) -> Result<Node, ParserError> {
        self.consume()?;
        self.next_node()
    }

    fn parse_script_tag_open(&mut self, s: &str) -> Result<Node, ParserError> {
        self.consume()?;

        let parts = s.split_whitespace();

        let mut attributes = BTreeMap::new();

        for attr in parts {
            if let Some((key, value)) = attr.split_once('=') {
                attributes.insert(
                    key.to_string(),
                    AttributeValue::Value(value.trim_matches('"').to_string()),
                );
            } else {
                attributes.insert(attr.to_string(), AttributeValue::Boolean);
            }
        }

        let mut children = Vec::new();

        while !self.is_at_end() {
            match self.next_node() {
                Ok(node) => {
                    children.push(node);
                }
                Err(ParserError::ClosingTagFound(tag)) => {
                    if tag == "script" {
                        self.consume()?;
                        break;
                    }
                    // If it's not our closing tag, keep collecting children
                }
                Err(e) => return Err(e),
            }
        }

        Ok(Node::Script(ScriptNode::Element {
            attributes,
            children,
        }))
    }

    fn parse_style_tag_open(&mut self, s: &str) -> Result<Node, ParserError> {
        self.consume()?;

        let mut parts = s.split_whitespace();

        let _tag_name = parts
            .next()
            .ok_or(ParserError::InvalidTokenAccess)?
            .to_string();

        let mut attributes = BTreeMap::new();

        for attr in parts {
            if let Some((key, value)) = attr.split_once('=') {
                attributes.insert(
                    key.to_string(),
                    AttributeValue::Value(value.trim_matches('"').to_string()),
                );
            } else {
                attributes.insert(attr.to_string(), AttributeValue::Boolean);
            }
        }

        let mut children = Vec::new();

        while !self.is_at_end() {
            match self.next_node() {
                Ok(node) => {
                    children.push(node);
                }
                Err(ParserError::ClosingTagFound(tag)) => {
                    if tag == "style" {
                        self.consume()?;
                        break;
                    }
                    // If it's not our closing tag, keep collecting children
                }
                Err(e) => return Err(e),
            }
        }

        Ok(Node::Style(StyleNode::Element {
            attributes,
            children,
        }))
    }

    fn parse_text(&mut self, s: &str) -> Result<Node, ParserError> {
        self.consume()?;

        Ok(Node::Text(s.to_string()))
    }

    fn parse_whitespace(&mut self) -> Result<Node, ParserError> {
        self.consume()?;
        self.next_node()
    }

    fn peek(&self) -> Result<Token, ParserError> {
        self.peek_at(0)
    }

    fn peek_next(&self) -> Result<Token, ParserError> {
        self.peek_at(1)
    }

    fn peek_previous(&self) -> Result<Token, ParserError> {
        self.peek_at(-1)
    }

    fn peek_forward(&self, steps: usize) -> Result<Vec<Token>, ParserError> {
        (0..steps).map(|i| self.peek_at(i as isize)).collect()
    }

    fn peek_back(&self, steps: usize) -> Result<Vec<Token>, ParserError> {
        (1..=steps).map(|i| self.peek_at(-(i as isize))).collect()
    }

    fn peek_at(&self, offset: isize) -> Result<Token, ParserError> {
        let index = self.current as isize + offset;
        self.item_at(index as usize)
    }

    fn item_at(&self, index: usize) -> Result<Token, ParserError> {
        if let Some(token) = self.tokens.get(index) {
            Ok(token.clone())
        } else {
            let error = if self.tokens.is_empty() {
                ParserError::EmptyTokenStream
            } else if index < self.current {
                ParserError::AtBeginningOfStream
            } else if index >= self.tokens.len() {
                ParserError::AtEndOfStream
            } else {
                ParserError::InvalidTokenAccess
            };
            Err(error)
        }
    }

    fn is_at_end(&self) -> bool {
        self.current + 1 >= self.tokens.len()
    }

    fn consume(&mut self) -> Result<Token, ParserError> {
        if self.is_at_end() {
            return Err(ParserError::AtEndOfStream);
        }
        self.current += 1;
        self.peek_previous()
    }

    fn backtrack(&mut self, steps: usize) -> Result<Token, ParserError> {
        if self.current < steps {
            return Err(ParserError::AtBeginningOfStream);
        }
        self.current -= steps;
        self.peek_next()
    }

    fn lookahead(&self, types: &[TokenType]) -> Result<bool, ParserError> {
        for (i, t) in types.iter().enumerate() {
            if !self.peek_at(i as isize)?.is_token_type(t) {
                return Ok(false);
            }
        }
        Ok(true)
    }

    fn consume_if(&mut self, token_type: TokenType) -> Result<Token, ParserError> {
        let token = self.peek()?;
        if token.is_token_type(&token_type) {
            return Err(ParserError::ExpectedTokenType(
                self.peek()?.clone(),
                token_type,
            ));
        }
        self.consume()?;
        Ok(token)
    }

    fn consume_until(&mut self, end_type: TokenType) -> Result<Vec<Token>, ParserError> {
        let mut consumed = Vec::new();
        while !self.is_at_end() && self.peek()?.is_token_type(&end_type) {
            let token = self.consume()?;
            consumed.push(token);
        }
        Ok(consumed)
    }

    fn synchronize(&mut self, sync_types: &[TokenType]) -> Result<(), ParserError> {
        while !self.is_at_end() {
            if sync_types.contains(self.peek()?.token_type()) {
                return Ok(());
            }
            self.consume()?;
        }
        Ok(())
    }
}

#[derive(Error, Debug)]
pub enum ParserError {
    #[error("token stream is empty")]
    EmptyTokenStream,
    #[error("at beginning of token stream")]
    AtBeginningOfStream,
    #[error("at end of token stream")]
    AtEndOfStream,
    #[error("invalid token access")]
    InvalidTokenAccess,
    #[error("unexpected token '{0:?}', expected type '{1:?}'")]
    ExpectedTokenType(Token, TokenType),
    #[error("unexpected token '{0:?}'")]
    UnexpectedToken(Token),
    #[error("unexpected end tag: {0}")]
    UnexpectedEndTag(String),
    #[error("multi-line comment outside of script or style context")]
    InvalidMultLineComment,
    #[error("unexpected end of file")]
    UnexpectedEof,
    #[error("found closing tag: {0}")]
    ClosingTagFound(String),
    #[error(transparent)]
    Node(#[from] AstError),
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;

    #[test]
    fn test_parse_comments() {
        let source = r#"<!-- HTML comment -->
{# Django comment #}
<script>
    // JS single line
    /* JS multi
        line */
</script>
<style>
    /* CSS comment */
</style>"#;
        let tokens = Lexer::new(source).tokenize().unwrap();
        let mut parser = Parser::new(tokens);
        let ast = parser.parse().unwrap();
        insta::assert_yaml_snapshot!(ast);
    }

    #[test]
    fn test_parse_django_block() {
        let source = r#"{% if user.is_staff %}Admin{% else %}User{% endif %}"#;
        let tokens = Lexer::new(source).tokenize().unwrap();
        let mut parser = Parser::new(tokens);
        let ast = parser.parse().unwrap();
        insta::assert_yaml_snapshot!(ast);
    }

    #[test]
    fn test_parse_django_variable() {
        let source = r#"{{ user.name|default:"Anonymous"|title }}"#;
        let tokens = Lexer::new(source).tokenize().unwrap();
        let mut parser = Parser::new(tokens);
        let ast = parser.parse().unwrap();
        insta::assert_yaml_snapshot!(ast);
    }
    #[test]
    fn test_parse_html_tag() {
        let source = r#"<div class="container" id="main" disabled></div>"#;
        let tokens = Lexer::new(source).tokenize().unwrap();
        let mut parser = Parser::new(tokens);
        let ast = parser.parse().unwrap();
        insta::assert_yaml_snapshot!(ast);
    }

    #[test]
    fn test_parse_script() {
        let source = r#"<script type="text/javascript">
    // Single line comment
    const x = 1;
    /* Multi-line
        comment */
    console.log(x);
</script>"#;
        let tokens = Lexer::new(source).tokenize().unwrap();
        let mut parser = Parser::new(tokens);
        let ast = parser.parse().unwrap();
        insta::assert_yaml_snapshot!(ast);
    }

    #[test]
    fn test_parse_style() {
        let source = r#"<style type="text/css">
    /* Header styles */
    .header {
        color: blue;
    }
</style>"#;
        let tokens = Lexer::new(source).tokenize().unwrap();
        let mut parser = Parser::new(tokens);
        let ast = parser.parse().unwrap();
        insta::assert_yaml_snapshot!(ast);
    }

    fn test_parse_full() {
        let source = r#"<!DOCTYPE html>
<html>
    <head>
        <style type="text/css">
            /* Style header */
            .header { color: blue; }
        </style>
        <script type="text/javascript">
            // Init app
            const app = {
                /* Config */
                debug: true
            };
        </script>
    </head>
    <body>
        <!-- Header section -->
        <div class="header" id="main" data-value="123" disabled>
            {% if user.is_authenticated %}
                {# Welcome message #}
                <h1>Welcome, {{ user.name|default:"Guest"|title }}!</h1>
                {% if user.is_staff %}
                    <span>Admin</span>
                {% else %}
                    <span>User</span>
                {% endif %}
            {% endif %}
        </div>
    </body>
</html>"#;
        let tokens = Lexer::new(source).tokenize().unwrap();
        let mut parser = Parser::new(tokens);
        let ast = parser.parse().unwrap();
        insta::assert_yaml_snapshot!(ast);
    }

    #[test]
    fn test_parse_unexpected_eof() {
        let source = "<div>\n";
        let tokens = Lexer::new(source).tokenize().unwrap();
        let mut parser = Parser::new(tokens);
        let ast = parser.parse();
        assert!(matches!(ast, Err(ParserError::UnexpectedEof)));
    }
}
