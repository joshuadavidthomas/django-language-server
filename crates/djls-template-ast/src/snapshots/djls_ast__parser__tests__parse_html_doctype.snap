---
source: crates/djls-ast/src/parser.rs
expression: ast
---
nodes:
  - Html:
      Doctype: "!DOCTYPE"
